Split enrollments of the ValidatorSet into candidates and validators

linked0 opened the pull request:
The old `ValidatorSet` had an issue that it manages two kinds of enrollments which consist of the enrollments waiting to be a validator and the enrollments having already been a validator. This could cause an unexpected error, so we are splitting the class into the `EnrollmentPoo`l for the candidates of a validator and the `ValidatorSet` for the confirmed validators.

**In this PR, I have just split the ValidatorSet class into two classes, EnrollmentPool and ValidatorSet for smaller PR.** 

There will be other PRs having the following modifications.
1. Change unnecessary public into private in the `ValidatorSet`
2. Change function names in the `ValidatorSet` into proper names like `getEnrollment` into `getValidator`
3. Check abnormal situations in `add` of the `ValidatorSet` and `updateEnrolledHeight` of the EnrollmentManager.

Relates #735

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,376 @@
+                this.db.execute("INSERT INTO enrollment_pool (key, val) VALUES (?, ?)",
+                    enroll.utxo_key.toString(), buffer);
+            }();
+
```
```suggestion
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,376 @@
+
+        Returns:
+            the number of enrollments being managed by the enrollment pool,
+            which includes both registered and un-registered enrollments.
```
In `EnrollmentManager.updateEnrolledHeight`, the enrollment is removed from the pool when the height is updated. I assume that when the height is updated it means the enrollment has been registered. So then the pool only contains unregistered enrollments, right?

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,376 @@
+
+        // create the table for enrollment pool if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS enrollment_pool " ~
+            "(key TEXT PRIMARY KEY, val BLOC NOT NULL)");
```
```suggestion
            "(key TEXT PRIMARY KEY, val BLOB NOT NULL)");
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -22,6 +22,7 @@ import agora.consensus.data.Block;
 import agora.consensus.data.Enrollment;
```
Typo in commit message: candicates => candidates

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -212,7 +218,13 @@ public class EnrollmentManager
+        if (!this.enroll_pool.getEnrollment(enroll_hash, enroll))
+            return false;
+
+        this.enroll_pool.remove(enroll_hash);
```
There is a problem here:

- We take one enrollment from the pool
- We remove it from the pool
- We try to add it to the validator set

If the last step fails the enrollment no longer exists anywhere because it was removed from the pool. We should only call `this.enroll_pool.remove(enroll_hash);` if updating the enrolled height (adding it to the validator set) was successful.

linked0 replied:
Thanks for your comment. In the case, the function returns `false` and the ledger will terminate a node with the `assert(0)` call. The validator set can be restored with the ledger when a node starts.

AndrejMitrovic replied:
Then the assert should be moved inside of here. Otherwise this is what's called "leaky abstractions".

linked0 replied:
It could be better for the ledger to have the role of determining whether a node terminates or not. But, I'll think about what you suggested.

AndrejMitrovic replied:
If you want the Ledger to have the option to recover, then you need to make sure the EnrollmentManager is in a valid state.

That means to not remove the enrollment from the pool unless it was successfully added to the validator set.

linked0 replied:
The information for validators is stored in a block and the ledger restores validators from the block. At the time when the `updateEnrolledHeight` is being called, the validators' information has already been stored in a block.

Geod24 replied:
> It could be better for the ledger to have the role of determining whether a node terminates or not.

Ideally, it would be better if the node would not terminate ;)

> At the time when the `updateEnrolledHeight` is being called, the validators' information has already been stored in a block.

Right, but we don't read the block here, we read the `enroll_pool`. And if the `enroll_pool` does not contain the enrollment, we bail out. No checking the block, no requesting other nodes.

Also, I didn't see where it was called from ?

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -83,56 +80,40 @@ public class ValidatorSet
+                    enroll);
+                return false;
+            }
+            else
```
No need for `else` if you `return`.

linked0 replied:
Please share your idea in more detail.

AndrejMitrovic replied:
```D
if (condition)
{
    do();
    return;
}
else
{
    something();
}
```

Is the same as:

```D
if (condition)
{
    do();
    return;
}

something();
```

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -26,71 +26,7 @@ import agora.consensus.EnrollmentManager;
 import agora.consensus.Genesis;
 import agora.test.Base;
 
-/// test enrollment process
```
Why was this removed ?

linked0 replied:
The unit test code has been merged into the code below the comment `/// test for  enrollment process & revealing a pre-image periodically`. Changing the way the EnrollmentManager stores the data make the test code need to be changed and I found there was redundant test code, I removed the redundant code and merged the part of it into the other test code.

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -788,7 +681,7 @@ unittest
     // make test blocks used for restoring validator set
     Block[] blocks;
-    ulong last_height = ValidatorSet.ValidatorCycle;
+    ulong last_height = Enrollment.ValidatorCycle;
```
There's a lot of diff coming from simply moving this constant. I think it should be its own commit.

linked0 replied:
I will do it. but in some commit of this PR, there will be two `ValidatorCycle` constants because the `EnrollmentPool` must reference it. In the end, the ValidatorCycle will only have existed in the `Enrollment` class.

Geod24 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,375 @@
+        this.db = Database(db_path);
+
+        // create the table for enrollment pool if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS enrollment_pool " ~
```
Do we really need a table here ?

linked0 replied:
If we don't need this new table, I didn't need to start to resolve this issue. This PR and the issue related to this is all about creating a new table like enrollment_pool.

linked0 replied:
> No checking the block, no requesting other nodes.

You're right. We need some recovery code like requesting enrollments from other nodes. We need another issue for that and I'm sorry for the late catch.

linked0 replied:
> No checking the block, no requesting other nodes

I have created an issue for the problem. #755

Geod24 reviewed (changes requested):
A few comments, let's discuss IRL the rest for simplicity.

linked0 commented:
Ready for review.
I resolved the comments and created a new issue for an issue that can't be handled in this PR.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -847,7 +859,6 @@ unittest
     assert(man.getEnrolledHeight(utxo_hash2) == 0);
     man.getUnregistered(enrolls);
     assert(enrolls.length == 1);
-    assert(man.count() == 2);  // has not changed
```
Why was the unittest removed instead of just changing the number that's tested?

linked0 replied:
> Why was the unittest removed instead of just changing the number that's tested?

This kind of deletion of code could be confusing. I'll restore it.

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,374 @@
+                enroll);
+            return false;
+        }
+
```
The `buffer` should be declared here where it's initially used, not at the start of the function. We always try to keep declarations close to where they are initially used, it prevents accidental usage of uninitialized variables.

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,374 @@
+mixin AddLogger!();
+
+/// Ditto
+public class EnrollmentPool
```
There is an inconsistency in the API.

`hasEnrollment` doesn't throw, but `count` throws, `add` doesn't throw, but `remove` throws, and then `getEnrollment` throws and **also** returns a status flag.

These methods need to be consistent.

linked0 replied:
> These methods need to be consistent.

Right, those functions should be consistent. I'll make it so.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -130,44 +135,45 @@ public class EnrollmentManager
         Returns:
-            the number of enrollments being managed by this EnrollmentManager.
+            the number of enrollments being managed by the enrollment pool.
             Note: this includes both registered and un-registered enrollments.
```
I think the Note: should be removed, because the pool only tracks un-registered enrollments.

linked0 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,374 @@
+        static ubyte[] buffer;
+
+        // check validity of the enrollment data
+        if (auto reason = isInvalidEnrollmentReason(enroll, block_height + 1,
+            finder))
```
The `block_height` is not used properly, the function of `isInvalidEnrollmentReason` should be changed.

linked0 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -83,56 +84,38 @@ public class ValidatorSet
-
         try
         {
+            if (this.hasEnrollment(enroll.utxo_key))
+            {
+                log.info("Rejected already existing validator, Data was: {}",
+                    enroll);
+                return false;
+            }
```
This code doesn't need to be here in this `try/catch` statement.

linked0 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -83,56 +84,38 @@ public class ValidatorSet
     {
         static ubyte[] buffer;
-
-        // check validity of the enrollment data
-        if (auto reason = isInvalidEnrollmentReason(enroll, block_height + 1,
-            finder))
-        {
-            log.info("Invalid enrollment data: {}, Data was: {}", reason, enroll);
-            return false;
-        }
```
Revert it

linked0 committed:
Split enrollments into candidates and validators

The ValidatorSet had an issue that it manages two kinds of enrollments
which consist of the enrollments waiting to be a validator and the
enrollments having already been a validator. This could cause an
unexpected error, so we split the class into the EnrollmentPool for
the candidates of a validator and the ValidatorSet for the confirmed
validators.

linked0 commented:
It needs some more modification. I'll request a review after the modifications are done.

linked0 commented:
Ready for review about the code that Mathias and I pair programmed.

Geod24 commented:
Could you put the first commit in its own PR ?

linked0 commented:
> Could you put the first commit in its own PR ?

OK. I will do it.

linked0 committed:
Remove the redundant constant, or ValidatorCycle from ValidatorSet

linked0 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,404 @@
+
+    ***************************************************************************/
+
+    public bool add (scope UTXOFinder finder, const ref Enrollment enroll)
```
Invert params

linked0 reviewed (commented):
Some suggestions.

TrustHenry commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -130,25 +135,26 @@ public class EnrollmentManager
 
         Returns:
-            true if the enrollment data has been added to the validator set
+            true if the enrollment data has been added to the the enrollment pool
```
```suggestion
            true if the enrollment data has been added to the enrollment pool
```

TrustHenry commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -310,7 +262,7 @@ public class ValidatorSet
             () @trusted {
                 auto results = this.db.execute("SELECT val FROM validator_set" ~
-                    " WHERE enrolled_height is not null");
+                    " ORDER BY key ASC");
```
> Get all the current validators

This change requires the document to be changed.
The meaning of sorted should be added.

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,404 @@
+
+    public size_t count () @trusted nothrow
+    {
+        size_t cnt;
+        try
+        {
+            cnt = this.db.execute("SELECT count(*) FROM enrollment_pool").
+                oneValue!size_t;
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error on count");
+        }
```
```suggestion
        try
        {
            return this.db.execute("SELECT count(*) FROM enrollment_pool").
                oneValue!size_t;
        }
        catch (Exception ex)
        {
            log.error("Database operation error on count");
            return 0;
        }
```

And the documentation needs to be updated to say 0 is returned in case of errors.

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -0,0 +1,404 @@
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error on remove");
```
When I said before that the API needs to be consistent I meant the code should do either one of the following:

A) Return a status code for every function which can fail. So the calling code can handle it.
B) Make all functions throw if they can fail. The calling code then must use a try/catch to handle it, or it can let the exception propagate upwards so other calling code handles it.
C) Catch any exceptions and halt the program with `assert(0)`.

Right now the code doesn't do any of these. Instead it catches the exception, logs it, and continues execution.

But what is the purpose of continuing execution like nothing bad happened? Will the database automatically fix itself on the next call? What is the expectation here?

linked0 replied:
The failure to remove an enrollment data from the enrollment pool is not so critical and so execution can continue. In this PR, if adding the enrollment to the validator set fails, the ledger makes the system terminate as follows.

```
foreach (enrollment; block.header.enrollments)
    if (!this.enroll_man.addValidator(enrollment,
            block.header.height, this.utxo_set.getUTXOFinder()))
        assert(0);
```

Mathias and I discussed that in the pair programming session.
The recovery from the database failure can be executed when a node start, which is not implemented yet and could be another issue.

linked0 commented:
Ready for review.
Additionally, I will make another PR about deleting the table named `node_enroll_data` used in the `EnrollmentManager` class soon like next Monday. The table manages only the `next_reveal_height` value which can be just set to the block height after the validator set is restored through the `restoreValidators` function.

Geod24 commented:
Rebased on v0.x.x
